### PR TITLE
feat(spanner): support request tagging

### PR DESCRIPTION
**Pull request by jiren**
Adds support for request options with request and transaction tags.

closes: #8162

### CODE REVIEW

File: google-cloud-spanner/lib/google/cloud/spanner/client.rb
Comment (skuruppu) [inline google-cloud-spanner/lib/google/cloud/spanner/client.rb:399]: What if this `execute_query` is part of a transaction? How does the transaction tag get passed in?

File: google-cloud-spanner/test/google/cloud/spanner/client/commit_transaction_tag_test.rb
Comment (skuruppu) [inline google-cloud-spanner/test/google/cloud/spanner/client/commit_transaction_tag_test.rb:1]:
```suggestion
# Copyright 2020 Google LLC
```

**Review by skuruppu — COMMENTED** [review]
Just a couple of minor comments.

File: google-cloud-spanner/lib/google/cloud/spanner/client.rb
Comment (jiren) [reply google-cloud-spanner/lib/google/cloud/spanner/client.rb]: Client can execute only single use transaction, so here only request tag will work. Transaction tag on `execute_query` is  works using transaction object. 

```ruby
db.transaction tag: "Tag-1" do |tx|
  results = tx.execute_query "SELECT * FROM users", tag: "Tag-1-1"
end
```

Implementation ref: https://github.com/googleapis/google-cloud-ruby/blob/880f64fe6db2d48ef989fad356a21d23550cbd88/google-cloud-spanner/lib/google/cloud/spanner/client.rb#L1352-L1373

Here on line 1364 `truncation_tag` will set and pass to transaction object to add queries executed by transaction.

File: google-cloud-spanner/lib/google/cloud/spanner/client.rb
Comment (jiren) [reply google-cloud-spanner/lib/google/cloud/spanner/client.rb]: Done

File: google-cloud-spanner/lib/google/cloud/spanner/client.rb
Comment (jiren) [reply google-cloud-spanner/lib/google/cloud/spanner/client.rb]: Done

File: google-cloud-spanner/lib/google/cloud/spanner/transaction.rb
Comment (jiren) [reply google-cloud-spanner/lib/google/cloud/spanner/transaction.rb]: Yes, moved to `Convert` helper file.

File: google-cloud-spanner/test/google/cloud/spanner/client/commit_transaction_tag_test.rb
Comment (jiren) [reply google-cloud-spanner/test/google/cloud/spanner/client/commit_transaction_tag_test.rb]: Done

File: google-cloud-spanner/lib/google/cloud/spanner/client.rb
Comment (jiren) [reply google-cloud-spanner/lib/google/cloud/spanner/client.rb]: Done

File: google-cloud-spanner/lib/google/cloud/spanner/client.rb
Comment (thiagotnunes) [reply google-cloud-spanner/lib/google/cloud/spanner/client.rb]: Sorry, just to verify I understood it correctly. On the example you provided, the transaction tag will be `Tag-1` and the request tag will be `Tag-1-1`, right?

File: google-cloud-spanner/lib/google/cloud/spanner/client.rb
Comment (thiagotnunes) [inline google-cloud-spanner/lib/google/cloud/spanner/client.rb:392]: I wonder if the tag should not be in the `query_options` or `call_options` hashmap, wdyt? Or is this a private method used internally only?

File: google-cloud-spanner/lib/google/cloud/spanner/client.rb
Comment (thiagotnunes) [inline google-cloud-spanner/lib/google/cloud/spanner/client.rb:1242]: The commit (transaction) `tag` should be defined in the transaction creation, like in your previous example:

```
db.transaction tag: "Tag-1" do |tx| ...
```

The above transaction tag should be included in the commit statement.

If this is for hand-crafted transactions, I still think we should define the transaction tag when the user creates the transaction initially.

Comment (thiagotnunes) [reply google-cloud-spanner/lib/google/cloud/spanner/client.rb:1242]: Or is this method used internally only?

File: google-cloud-spanner/lib/google/cloud/spanner/client.rb
Comment (jiren) [reply google-cloud-spanner/lib/google/cloud/spanner/client.rb]: Yes "tag-1" is transaction tag and "Tag-1-1" is request tag.

File: google-cloud-spanner/lib/google/cloud/spanner/client.rb
Comment (jiren) [reply google-cloud-spanner/lib/google/cloud/spanner/client.rb:392]: No, this is not a private method. In my view, we should not use `query_options` or `call_options`.`query_options` for optimization params and `call_options` is not for apis it is for GRPC client. We can use `tag` as `request_options` as per [proto file](https://github.com/googleapis/googleapis/blob/0dd48adfa0aa8f4d042813af248dcbeecc5f7102/google/spanner/v1/spanner.proto#L566) but client design specs "tag" is used. i.e ```roTx.read(table, keySet, columns, Tag.of("tag-2"));``` 

@thiagotnunes  Let me know your thoughts.

File: google-cloud-spanner/lib/google/cloud/spanner/client.rb
Comment (jiren) [reply google-cloud-spanner/lib/google/cloud/spanner/client.rb:1242]: `commit` is public method. 

For hand-crafted transaction :
https://github.com/googleapis/google-cloud-ruby/blob/880f64fe6db2d48ef989fad356a21d23550cbd88/google-cloud-spanner/lib/google/cloud/spanner/client.rb#L1352-L1373

In this 'Tag-1' set to a transaction(L1364) which is use to set a transaction tag in API query calls executed on that transaction.

File: google-cloud-spanner/lib/google/cloud/spanner/client.rb
Comment (thiagotnunes) [reply google-cloud-spanner/lib/google/cloud/spanner/client.rb:392]: I think it would be nice to define a `request_options`, because that will make the API more extensible (we have other features coming soon, for which we could piggy back on this new hashmap then).

For the client design spec, we have the `roTx.read(table, keySet, columns, Tag.of("tag-2"));`, but if you check the implementation of that, you can see that `Tag` is actually a flavour of `Option` as well (https://github.com/googleapis/java-spanner/pull/576/files#diff-992d480774d874d97423da09934b0395ef460130558e2fd6f7b1461ba32535c0R154-R166). So, if we have the `request_options` map here, we would not be deviating from the original design.

File: google-cloud-spanner/lib/google/cloud/spanner/client.rb
Comment (thiagotnunes) [reply google-cloud-spanner/lib/google/cloud/spanner/client.rb:1242]: Got it, so do we need the tag in the `commit` method? I think we should remove that, since it should be set during the transaction creation, right? Otherwise, what would happen if the user sets a transaction tag in the creation and another one in the commit?

File: google-cloud-spanner/lib/google/cloud/spanner/client.rb
Comment (jiren) [reply google-cloud-spanner/lib/google/cloud/spanner/client.rb:392]: @thiagotnunes Thanks for detailed information. Making changes according  to it.

File: google-cloud-spanner/lib/google/cloud/spanner/client.rb
Comment (jiren) [reply google-cloud-spanner/lib/google/cloud/spanner/client.rb:1242]: I think no need to remove it. There is a commit method in the client so the user might directly call it, while Transaction calls the internal private service object commit method. 

Client : user can directly use.
https://github.com/googleapis/google-cloud-ruby/blob/bf17920882c3297f66d2249cd023bb2883e707b6/google-cloud-spanner/lib/google/cloud/spanner/client.rb#L1095

Service: Private 
https://github.com/googleapis/google-cloud-ruby/blob/bf17920882c3297f66d2249cd023bb2883e707b6/google-cloud-spanner/lib/google/cloud/spanner/service.rb#L399

File: google-cloud-spanner/lib/google/cloud/spanner/client.rb
Comment (jiren) [reply google-cloud-spanner/lib/google/cloud/spanner/client.rb:392]: @thiagotnunes  So method will look like this. Please confirm.

```
 def execute_query sql, params: nil, types: nil, single_use: nil,
                                  query_options: nil, call_options: nil
                                  query_options: nil, request_options: nil,
```
Use:
```
client.execute_query("SELECT ...", request_options: { request_tag: "Tag-1" })
```

For transaction:
```
def transaction deadline: 120, request_options: nil, call_options: nil
```
Use
```
  client.transaction request_options: { transaction_tag: "Tag-1 } do 
  end
```

File: google-cloud-spanner/lib/google/cloud/spanner/client.rb
Comment (skuruppu) [reply google-cloud-spanner/lib/google/cloud/spanner/client.rb]: Thanks for clarifying.

File: google-cloud-spanner/lib/google/cloud/spanner/client.rb
Comment (skuruppu) [reply google-cloud-spanner/lib/google/cloud/spanner/client.rb:392]: I think this is a reasonable design. When you implement it, just make sure that in the documentation for the function, it is clear if the other type of tag is ignored. Something like "if transaction_tag is specified, it will be ignored".

Comment (jiren) [reply google-cloud-spanner/lib/google/cloud/spanner/client.rb:392]: ok. @skuruppu Do we need to implement ignore tag logic or just documentation is sufficient?

**Review by skuruppu — APPROVED** [review]

File: google-cloud-spanner/lib/google/cloud/spanner/client.rb
Comment (thiagotnunes) [reply google-cloud-spanner/lib/google/cloud/spanner/client.rb]: @jiren I think that the user should never have to decide whether the tag should be a `request_tag` or `transaction_tag`, but that should be done internally. We could change the example to:

```
client.transaction request_options: { tag: "Tag-1 } do 
end
```

```
client.execute_query("SELECT ...", request_options: { tag: "Tag-1" })
```

Wdyt?

Comment (jiren) [reply google-cloud-spanner/lib/google/cloud/spanner/client.rb]: @thiagotnunes yes this will be helpful for users. I will start making changes for this.

Comment (jiren) [reply google-cloud-spanner/lib/google/cloud/spanner/client.rb]: @thiagotnunes Updated as per above examples.

File: google-cloud-spanner/lib/google/cloud/spanner/client.rb
Comment (thiagotnunes) [inline google-cloud-spanner/lib/google/cloud/spanner/client.rb:396]: Just realised that we added the `request_options` as the sixth parameter, where that position used to be used by the `call_options`. The type of both parameters is a map, which might be silently fail for users who had previous code written and were not naming the arguments.

Do you think it should be worth to move it to be the last argument of the method?
Also, could you make sure that this PR is a breaking change, since it updates this interface? (You can do that through the commit message, see https://www.conventionalcommits.org/en/v1.0.0/#specification BREAKING CHANGE).

Comment (jiren) [reply google-cloud-spanner/lib/google/cloud/spanner/client.rb:396]: These are named arguments with the default value. In Ruby named arguments are like hash so, the order has no impact while calling.  Here `call_options` are for the GAPIC GRPC client so moved to last, to separate from spanner arguments parameters.

Comment (thiagotnunes) [reply google-cloud-spanner/lib/google/cloud/spanner/client.rb:396]: Makes sense, thanks for the explanation

**Review by thiagotnunes — APPROVED** [review]

**Comment by larkee** [comment]
@jiren Please rebase this

**Comment by jiren** [comment]
> @jiren Please rebase this

@thiagotnunes Updated code.

**Comment by skuruppu** [comment]
@jiren will you be able to resolve the conflicts?

@hengfengli I think we need your approval here before the PR can be merged.

**Review by hengfengli — APPROVED** [review]

**Comment by jiren** [comment]
> @jiren will you be able to resolve the conflicts?
> 
> @hengfengli I think we need your approval here before the PR can be merged.

@skuruppu rebased and resolved conflicts.

**Pull request merged**
